### PR TITLE
Ensure strings in decorator

### LINKED ISSUE #40937: influxdb_return.py throws exception
### Description of Issue/Question
Attempting to use --return influxdb with test.ping results in an exception in the minion log, and no data written to influxdb
### Setup
```bash
# cat /etc/salt/master | grep influxdb
influxdb.host: 'influxdb'
influxdb.db: 'metrics'
influxdb.user: 'writer'
influxdb.password: 'writer'
influxdb.port: 8086
```

### Steps to Reproduce Issue
```bash
# salt --return influxdb master test.ping
master:
    True
```
```python
2017-04-27 19:35:06,141 [salt.minion      ][INFO    ][23662] User root Executing command test.ping with jid 20170427193506130542
2017-04-27 19:35:06,141 [salt.minion      ][DEBUG   ][23662] Command details {'tgt_type': 'glob', 'jid': '20170427193506130542', 'tgt': 'master', 'ret': 'influxdb', 'user': 'root', 'arg': [], 'fun': 'test.ping'}
2017-04-27 19:35:06,157 [salt.minion      ][INFO    ][28564] Starting a new job with PID 28564
2017-04-27 19:35:06,181 [salt.utils.lazy  ][DEBUG   ][28564] LazyLoaded test.ping
2017-04-27 19:35:06,182 [salt.utils.lazy  ][DEBUG   ][28564] LazyLoaded direct_call.get
2017-04-27 19:35:06,182 [salt.minion      ][DEBUG   ][28564] Minion return retry timer set to 9 seconds (randomized)
2017-04-27 19:35:06,182 [salt.minion      ][INFO    ][28564] Returning information for job: 20170427193506130542
2017-04-27 19:35:06,183 [salt.transport.zeromq][DEBUG   ][28564] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', 'master', 'tcp://127.0.0.1:4506', 'aes')
2017-04-27 19:35:06,183 [salt.crypt       ][DEBUG   ][28564] Initializing new AsyncAuth for ('/etc/salt/pki/minion', 'master', 'tcp://127.0.0.1:4506')
2017-04-27 19:35:06,209 [salt.utils.lazy  ][DEBUG   ][28564] LazyLoaded influxdb.returner
2017-04-27 19:35:06,210 [salt.minion      ][ERROR   ][28564] The return failed for job 20170427193506130542 sequence item 1: expected string, int found
2017-04-27 19:35:06,210 [salt.minion      ][ERROR   ][28564] Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 1556, in _thread_return
    minion_instance.returners[returner_str](ret)
  File "/usr/lib/python2.7/dist-packages/salt/returners/influxdb_return.py", line 152, in returner
    serv = _get_serv(ret)
  File "/usr/lib/python2.7/dist-packages/salt/returners/influxdb_return.py", line 130, in _get_serv
    version = _get_version(host, port, user, password)
  File "/usr/lib/python2.7/dist-packages/salt/utils/decorators/__init__.py", line 248, in _memoize
    args_ = ','.join(list(args) + ['{0}={1}'.format(k, kwargs[k]) for k in sorted(kwargs)])
TypeError: sequence item 1: expected string, int found
```

### Versions Report
```
Salt Version:
           Salt: 2017.5.0-2-gfa2c8f8

Dependency Versions:
           cffi: 1.10.0
       cherrypy: Not Installed
       dateutil: 2.6.0
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: 1.2.3
      pycparser: 2.17
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.6 (default, Oct 26 2016, 20:30:19)
   python-gnupg: Not Installed
         PyYAML: 3.10
          PyZMQ: 14.0.1
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5

System Versions:
           dist: Ubuntu 14.04 trusty
        machine: x86_64
        release: 3.13.0-112-generic
         system: Linux
        version: Ubuntu 14.04 trusty

```

### PR DESCRIPTION
**Pull request by cachedout**
Fixes #40937

### CODE REVIEW

File: salt/utils/decorators/__init__.py [hunk 1]
```diff
@@ -245,7 +245,7 @@ def memoize(func):
 
     @wraps(func)
     def _memoize(*args, **kwargs):
-        args_ = ','.join(list(args) + ['{0}={1}'.format(k, kwargs[k]) for k in sorted(kwargs)])
+        args_ = ','.join([str(arg) for arg in args] + ['{0}={1}'.format(str(k), str(kwargs[k])) for k in sorted(kwargs)])
```
Comment (terminalmage) [inline salt/utils/decorators/__init__.py]: The problem with just using `str()` in PY2 is that it causes a ``UnicodeDecodeError`` when you do it on a unicode string:

```python
>>> x = [1, 2, 3, u'✓']
>>> x
[1, 2, 3, u'\u2713']
>>> y = [str(z) for z in x]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeEncodeError: 'ascii' codec can't encode character u'\u2713' in position 0: ordinal not in range(128)
```

So, while your code is more concise, an approach like [this](https://github.com/saltstack/salt/blob/c7c5266/salt/utils/timed_subprocess.py#L43-L49) would be better here.

**Review by terminalmage — CHANGES_REQUESTED** [review]

**Comment by mamalos** [comment]
@terminalmage 's comment is right. Thanks for the suggested solution.

**Comment by rallytime** [comment]
I have switched the base branch on this pull request from `nitrogen` to `2017.7` in preparation for moving from a code-named release to a version number.

File: salt/utils/decorators/__init__.py (hunk 1 above)
Comment (cachedout) [inline salt/utils/decorators/__init__.py]: @terminalmage Updates made.

**Review by terminalmage — APPROVED** [review]
:+1:

**Comment by mamalos** [comment]
Hi everybody!

Unfortunately this fix is buggy...first of all it broke some code I wrote (which means that I did not do my work well when you asked about my opinion before changing it :) ). But thanks to this bug I realised that I shouldn't have used the **memoize** decorator in the first place, so it turned out to be OK! :)

The first bug is in https://github.com/saltstack/salt/blob/2017.7.1/salt/utils/decorators/__init__.py#L254 (which for some reason is not present in **2017.7**), where we see that **args** is replaced with **str_args**, for some reason, which means that the actual arguments' list is changed (every argument is changed with its "string version") when this decorator is called. This, I suppose, shouldn't have been done (arguments should remain untouched). **2017.7** was correct wrt this problem.

The second bug that I see (and that it is based on the above discussion) is that the current code can hash wrong stuff. So, for example, a string argument (or keyword argument) with the value "False" will be cached over a boolean argument with the value False. So in order to be precise, the type of each argument should be added in the caching key as well. Hence, something like the following could be used instead:

```python
def _memoize(*args, **kwargs):
    args_ = ','.join(['{0}({1})'.format(x, type(x)) for x in args] + ['{0}={1}({2})'.format(k, kwargs[k], type(kwargs[k])) for k in sorted(kwargs)])
    if args_ not in cache:
        cache[args_] = func(*args, **kwargs)
    return cache[args_]
``` 
the **type** function could be any suggested type function (from module six or whatever), and I've removed the **isinstance** string conditionals since I'm not sure how they could help with the suggested string syntax. Of course, the above code is just my 2c, but I think you get the message.

**Comment by mamalos** [comment]
Yep, confirmed, @cro had already corrected this arguments' change in commit a260e91 for version **2017.7**. This should be merged in **2017.7.1** as well.

**Pull request merged**